### ISSUE TITLE
reqwest introduces a lot of transitive dependencies

### ISSUE THREAD

**Issue by danieldk** [jerry73204/rust-tfrecord#1]
First of all, thanks a lot for this crate! We use it in [sticker2](https://github.com/stickeritis/sticker2) to write Tensorboard summaries during training and it works great :tada: . However, adding `rust-tfrecord` as a dependency added *a lot* of transitive dependencies. The main culprit is `reqwest`:

```
│   ├── reqwest v0.10.8
│   │   ├── base64 v0.12.3
│   │   ├── bytes v0.5.6
│   │   ├── encoding_rs v0.8.24
│   │   │   └── cfg-if v0.1.10
│   │   ├── futures-core v0.3.5
│   │   ├── futures-util v0.3.5
│   │   │   ├── futures-core v0.3.5
│   │   │   ├── futures-io v0.3.5
│   │   │   ├── futures-task v0.3.5
│   │   │   │   └── once_cell v1.4.1
│   │   │   ├── memchr v2.3.3
│   │   │   ├── pin-project v0.4.23
│   │   │   │   └── pin-project-internal v0.4.23
│   │   │   │       ├── proc-macro2 v1.0.13 (*)
│   │   │   │       ├── quote v1.0.6 (*)
│   │   │   │       └── syn v1.0.22 (*)
│   │   │   ├── pin-utils v0.1.0
│   │   │   └── slab v0.4.2
│   │   ├── http v0.2.1
│   │   │   ├── bytes v0.5.6
│   │   │   ├── fnv v1.0.7
│   │   │   └── itoa v0.4.5
│   │   ├── http-body v0.3.1
│   │   │   ├── bytes v0.5.6
│   │   │   └── http v0.2.1 (*)
│   │   ├── hyper v0.13.7
│   │   │   ├── bytes v0.5.6
│   │   │   ├── futures-channel v0.3.5
│   │   │   │   └── futures-core v0.3.5
│   │   │   ├── futures-core v0.3.5
│   │   │   ├── futures-util v0.3.5 (*)
│   │   │   ├── h2 v0.2.6
│   │   │   │   ├── bytes v0.5.6
│   │   │   │   ├── fnv v1.0.7
│   │   │   │   ├── futures-core v0.3.5
│   │   │   │   ├── futures-sink v0.3.5
│   │   │   │   ├── futures-util v0.3.5 (*)
│   │   │   │   ├── http v0.2.1 (*)
│   │   │   │   ├── indexmap v1.3.2 (*)
│   │   │   │   ├── slab v0.4.2
│   │   │   │   ├── tokio v0.2.22
│   │   │   │   │   ├── bytes v0.5.6
│   │   │   │   │   ├── fnv v1.0.7
│   │   │   │   │   ├── futures-core v0.3.5
│   │   │   │   │   ├── iovec v0.1.4
│   │   │   │   │   │   └── libc v0.2.70
│   │   │   │   │   ├── lazy_static v1.4.0
│   │   │   │   │   ├── memchr v2.3.3
│   │   │   │   │   ├── mio v0.6.22
│   │   │   │   │   │   ├── cfg-if v0.1.10
│   │   │   │   │   │   ├── iovec v0.1.4 (*)
│   │   │   │   │   │   ├── libc v0.2.70
│   │   │   │   │   │   ├── log v0.4.8 (*)
│   │   │   │   │   │   ├── net2 v0.2.34
│   │   │   │   │   │   │   ├── cfg-if v0.1.10
│   │   │   │   │   │   │   └── libc v0.2.70
│   │   │   │   │   │   └── slab v0.4.2
│   │   │   │   │   ├── num_cpus v1.13.0 (*)
│   │   │   │   │   ├── pin-project-lite v0.1.7
│   │   │   │   │   └── slab v0.4.2
│   │   │   │   ├── tokio-util v0.3.1
│   │   │   │   │   ├── bytes v0.5.6
│   │   │   │   │   ├── futures-core v0.3.5
│   │   │   │   │   ├── futures-sink v0.3.5
│   │   │   │   │   ├── log v0.4.8 (*)
│   │   │   │   │   ├── pin-project-lite v0.1.7
│   │   │   │   │   └── tokio v0.2.22 (*)
│   │   │   │   └── tracing v0.1.19
│   │   │   │       ├── cfg-if v0.1.10
│   │   │   │       ├── log v0.4.8 (*)
│   │   │   │       └── tracing-core v0.1.15
│   │   │   │           └── lazy_static v1.4.0
│   │   │   ├── http v0.2.1 (*)
│   │   │   ├── http-body v0.3.1 (*)
│   │   │   ├── httparse v1.3.4
│   │   │   ├── itoa v0.4.5
│   │   │   ├── pin-project v0.4.23 (*)
│   │   │   ├── socket2 v0.3.12 (*)
│   │   │   ├── time v0.1.43 (*)
│   │   │   ├── tokio v0.2.22 (*)
│   │   │   ├── tower-service v0.3.0
│   │   │   ├── tracing v0.1.19 (*)
│   │   │   └── want v0.3.0
│   │   │       ├── log v0.4.8 (*)
│   │   │       └── try-lock v0.2.3
│   │   ├── hyper-tls v0.4.3
│   │   │   ├── bytes v0.5.6
│   │   │   ├── hyper v0.13.7 (*)
│   │   │   ├── native-tls v0.2.4
│   │   │   │   ├── log v0.4.8 (*)
│   │   │   │   ├── openssl v0.10.30
│   │   │   │   │   ├── bitflags v1.2.1
│   │   │   │   │   ├── cfg-if v0.1.10
│   │   │   │   │   ├── foreign-types v0.3.2
│   │   │   │   │   │   └── foreign-types-shared v0.1.1
│   │   │   │   │   ├── lazy_static v1.4.0
│   │   │   │   │   ├── libc v0.2.70
│   │   │   │   │   └── openssl-sys v0.9.58 (*)
│   │   │   │   ├── openssl-probe v0.1.2
│   │   │   │   └── openssl-sys v0.9.58 (*)
│   │   │   ├── tokio v0.2.22 (*)
│   │   │   └── tokio-tls v0.3.1
│   │   │       ├── native-tls v0.2.4 (*)
│   │   │       └── tokio v0.2.22 (*)
│   │   ├── ipnet v2.3.0
│   │   ├── lazy_static v1.4.0
│   │   ├── log v0.4.8 (*)
│   │   ├── mime v0.3.16
│   │   ├── mime_guess v2.0.3
│   │   │   ├── mime v0.3.16
│   │   │   └── unicase v2.6.0
│   │   │       [build-dependencies]
│   │   │       └── version_check v0.9.2
│   │   │   [build-dependencies]
│   │   │   └── unicase v2.6.0 (*)
│   │   ├── native-tls v0.2.4 (*)
│   │   ├── percent-encoding v2.1.0
│   │   ├── pin-project-lite v0.1.7
│   │   ├── serde v1.0.110 (*)
│   │   ├── serde_urlencoded v0.6.1
│   │   │   ├── dtoa v0.4.5
│   │   │   ├── itoa v0.4.5
│   │   │   ├── serde v1.0.110 (*)
│   │   │   └── url v2.1.1
│   │   │       ├── idna v0.2.0
│   │   │       │   ├── matches v0.1.8
│   │   │       │   ├── unicode-bidi v0.3.4
│   │   │       │   │   └── matches v0.1.8
│   │   │       │   └── unicode-normalization v0.1.12 (*)
│   │   │       ├── matches v0.1.8
│   │   │       └── percent-encoding v2.1.0
│   │   ├── tokio v0.2.22 (*)
│   │   ├── tokio-tls v0.3.1 (*)
│   │   └── url v2.1.1 (*)
```

Of course, there are duplicates, but it's still many additional dependencies. From `git grep`-ing, it seems that reqwest is only used for blocking `GET` requests. Would you be opposed to replacing these by the `curl` crate, which would reduce the number of dependencies a lot? For downstream crates it would already help a lot if reqwest was removed from `build-dependencies`/`build.rs`, because `dev-dependencies` are not used IIRC.

I can do a PR if this would be ok for you.

**Comment by danieldk** [comment]
Small update, by using `curl` in `build-dependencies`, the number of transitive dependencies of sticker2 changes from *292* to *230*.

**Comment by jerry73204** [comment]
I agree some deps are not necessary and it's definitely a large bloating. `curl` looks good to me except it introduces `curl-sys` that might breaks the compatibility on non-UNIX land. I'm looking at `ureq` to see if it can help.

**Comment by jerry73204** [comment]
Now `ureq` is shipped to v0.4.1. Please see if it works for you.

**Comment by danieldk** [comment]
Thanks a lot, that really helped. Down from *292* to *248*.
